From 15d63f58bf258894607eb1d346d8a8cc1fca55ea Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Fri, 24 Nov 2023 14:26:56 +0800
Subject: [PATCH] rss/kronos: Provision AP metadata and parse FIP image

Provision AP metadata to AP secure flash at AP preload
stage, as the AP secure flash can not be accessed before
SYSTOP is on.

Signed AP BL2 is in FIP placed in AP secure flash, so
parse the signed AP BL2 image from FIP then load it
to the SRAM of AP by RSS. The fip parser function
is referenced from 'corstone1000/fip_parser/fip_parser.c'.

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
---
 .../rss/common/bl1/bl1_1_shared_symbols.txt   |  1 +
 .../arm/rss/common/fip_parser/fip_parser.c    | 45 ++++++++++++
 .../arm/rss/common/fip_parser/fip_parser.h    | 38 ++++++++++
 .../ext/target/arm/rss/kronos/CMakeLists.txt  |  3 +
 .../target/arm/rss/kronos/bl2/boot_hal_bl2.c  | 66 ++++++++++++++++++
 .../target/arm/rss/kronos/bl2/flash_map_bl2.c | 18 ++---
 .../ext/target/arm/rss/kronos/flash_layout.h  | 68 +++++++++---------
 .../rss/kronos/fw_update_agent/fwu_agent.c    | 69 +++++++++++++++++++
 .../rss/kronos/fw_update_agent/fwu_agent.h    |  9 +++
 .../kronos/fw_update_agent/fwu_agent_rss.c    |  5 ++
 .../ext/target/arm/rss/kronos/size_defs.h     |  1 +
 11 files changed, 284 insertions(+), 39 deletions(-)

diff --git a/platform/ext/target/arm/rss/common/bl1/bl1_1_shared_symbols.txt b/platform/ext/target/arm/rss/common/bl1/bl1_1_shared_symbols.txt
index 312949e9f..6b989d7c1 100644
--- a/platform/ext/target/arm/rss/common/bl1/bl1_1_shared_symbols.txt
+++ b/platform/ext/target/arm/rss/common/bl1/bl1_1_shared_symbols.txt
@@ -1,5 +1,6 @@
 ATU_DEV_S
 Driver_FLASH0
+Driver_FLASH1
 KMU_DEV_S
 SystemInit
 atu_initialize_region
diff --git a/platform/ext/target/arm/rss/common/fip_parser/fip_parser.c b/platform/ext/target/arm/rss/common/fip_parser/fip_parser.c
index dd8fbc55b..95e6b0307 100644
--- a/platform/ext/target/arm/rss/common/fip_parser/fip_parser.c
+++ b/platform/ext/target/arm/rss/common/fip_parser/fip_parser.c
@@ -11,6 +11,9 @@
 #include "Driver_Flash.h"

 #include <string.h>
+#ifdef TFM_FWU_AGENT
+#include "bootutil/bootutil_log.h"
+#endif /* TFM_FWU_AGENT */

 extern ARM_DRIVER_FLASH FLASH_DEV_NAME;

@@ -84,3 +87,45 @@ int fip_get_entry_by_uuid(uint32_t fip_base, uint32_t atu_slot_size, uuid_t uuid
     /* UUID not found, return error. */
     return 3;
 }
+
+#ifdef TFM_FWU_AGENT
+int parse_fip_and_extract_tfa_info(uint32_t address, uint32_t fip_size,
+        uint32_t *tfa_offset, uint32_t *tfa_size)
+{
+    FIP_TOC_HEADER *toc_header = NULL;
+    FIP_TOC_ENTRY *toc_entry = NULL;
+    uuid_t uuid_null = {0};
+    uuid_t tfa_uuid = UUID_TRUSTED_BOOT_FIRMWARE_BL2;
+    uint8_t *iter;
+
+    toc_header = (FIP_TOC_HEADER *) address;
+
+    volatile uint32_t *tmp = address;
+
+    if (toc_header->name != TOC_HEADER_NAME) {
+        return FIP_PARSER_ERROR;
+    }
+
+    toc_entry = (FIP_TOC_ENTRY *)(toc_header + 1);
+
+    for (iter = (char *)toc_entry;
+        iter <= (char *)toc_header + fip_size;
+        iter = iter + sizeof(FIP_TOC_ENTRY), toc_entry++) {
+
+        if (!memcmp(iter, &uuid_null, sizeof(uuid_t))) {
+            return FIP_PARSER_ERROR;
+        }
+
+        if (!memcmp(iter, &tfa_uuid, sizeof(uuid_t))) {
+            BOOT_LOG_INF("TF-A FIP at : address = 0x%x : size = 0x%x \n\r",
+                                toc_entry->address,
+                                toc_entry->size);
+            *tfa_offset = toc_entry->address;
+            *tfa_size = toc_entry->size;
+            return FIP_PARSER_SUCCESS;
+        }
+    }
+
+    return FIP_PARSER_ERROR;
+}
+#endif /* TFM_FWU_AGENT */
diff --git a/platform/ext/target/arm/rss/common/fip_parser/fip_parser.h b/platform/ext/target/arm/rss/common/fip_parser/fip_parser.h
index b613a46aa..f61d98449 100644
--- a/platform/ext/target/arm/rss/common/fip_parser/fip_parser.h
+++ b/platform/ext/target/arm/rss/common/fip_parser/fip_parser.h
@@ -15,6 +15,44 @@
 extern "C" {
 #endif

+#include <stdint.h>
+#include "uuid.h"
+
+/* Return code of fip parser APIs */
+#define FIP_PARSER_SUCCESS                     (0)
+#define FIP_PARSER_ERROR                       (INT32_MIN + 1)
+
+/* This is used as a signature to validate the blob header */
+#define TOC_HEADER_NAME 0xAA640001
+
+/* ToC Entry UUIDs */
+#define UUID_TRUSTED_BOOT_FIRMWARE_BL2 \
+    {{0x5f, 0xf9, 0xec, 0x0b}, {0x4d, 0x22}, \
+    {0x3e, 0x4d}, 0xa5, 0x44, \
+    {0xc3, 0x9d, 0x81, 0xc7, 0x3f, 0x0a} }
+
+typedef struct _FIP_TOC_HEADER {
+    uint32_t    name;
+    uint32_t    serial_number;
+    uint64_t    flags;
+} FIP_TOC_HEADER;
+
+/* FIP tool creates fip table using 64 bit integers because of its
+ * basic support for 64 bit Cortex A.Hence need to add 32 bit padding
+ * for 32-bit Cortex-M CPUs
+ */
+
+typedef struct _FIP_TOC_ENTRY {
+    uuid_t      uuid;
+    uint32_t    address;
+    uint32_t    pad1;
+    uint32_t    size;
+    uint32_t    pad2;
+    uint64_t    flags;
+} FIP_TOC_ENTRY;
+
+int parse_fip_and_extract_tfa_info(uint32_t address, uint32_t fip_size,
+        uint32_t *tfa_offset, uint32_t *tfa_size);
 /**
  * \brief                    Parse a FIP and retrieve the offset and size of one
  *                           of the firmware images (specified by UUID).
diff --git a/platform/ext/target/arm/rss/kronos/CMakeLists.txt b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
index 1f954391a..aa50159a6 100644
--- a/platform/ext/target/arm/rss/kronos/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
@@ -112,6 +112,7 @@ target_sources(platform_bl2
         host_drivers/cfi_drv.c
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_atu_request_handler.c
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_mailbox.c
+        ${CMAKE_CURRENT_LIST_DIR}/../common/fip_parser/fip_parser.c
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011/uart_pl011_drv.c
         tfm_interrupts.c
         tower_nci_lib.c
@@ -133,6 +134,7 @@ target_compile_definitions(platform_bl2
     PRIVATE
         $<$<BOOL:${ENABLE_SCP_ATU_CTRL}>:ENABLE_SCP_ATU_CTRL>
         MEASURED_BOOT_HASH_ALG=${MEASURED_BOOT_HASH_ALG}
+        $<$<BOOL:${TFM_FWU_AGENT}>:TFM_FWU_AGENT>
 )

 target_sources(bl2
@@ -151,6 +153,7 @@ target_include_directories(platform_bl2
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
+        ${PLATFORM_DIR}/ext/target/arm/rss/common/fip_parser
         io
         soft_crc
         partition
diff --git a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
index 33d200082..7326b214a 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
@@ -37,6 +37,7 @@
 #include "platform.h"
 #include "partition.h"
 #include "fwu_agent.h"
+#include "fip_parser.h"

 extern volatile bool scp_doorbell;
 extern ARM_DRIVER_FLASH FLASH_DEV_NAME;
@@ -100,6 +101,59 @@ static bool fill_rss_flash_map_with_data(uint8_t boot_index) {
     return true;
 }

+static bool fill_secure_flash_map_with_data(uint8_t boot_index) {
+    uint32_t tfa_offset = 0;
+    size_t tfa_size = 0;
+    uint32_t fip_offset = 0;
+    size_t fip_size = 0;
+    int result;
+    uint8_t i = 0;
+    uint8_t id = 0;
+    uint8_t flash_id = 0;
+    partition_entry_t *entry = NULL;
+
+    if (boot_index >= NR_OF_FW_BANKS) {
+        BOOT_LOG_ERR("%d is an invalid boot_index, 0 <= boot_index < %d",
+                     boot_index, NR_OF_FW_BANKS);
+        return false;
+    }
+
+    partition_entry_t *fip_entry =
+        get_partition_entry(image_names[RSS_FIRMWARE_AP_BL2_ID][boot_index]);
+    if (fip_entry == NULL) {
+        BOOT_LOG_ERR("Could not find partition %s", image_names[RSS_FIRMWARE_AP_BL2_ID][boot_index]);
+        return false;
+    }
+
+    fip_offset = fip_entry->start;
+    fip_size = fip_entry->length;
+
+    result = parse_fip_and_extract_tfa_info (
+                     AP_FLASH_LOG_BASE + fip_offset + FIP_SIGNATURE_AREA_SIZE,
+                     fip_size, &tfa_offset, &tfa_size);
+    if (result != FIP_PARSER_SUCCESS) {
+        BOOT_LOG_ERR("parse_fip_and_extract_tfa_info failed");
+        return false;
+    }
+
+    for (i = 0; i < flash_map_entry_num; i++) {
+        id = flash_map[i].fa_id;
+
+        if(boot_index == BANK_0) {
+            flash_id = FLASH_AREA_IMAGE_PRIMARY(RSS_FIRMWARE_AP_BL2_ID);
+        } else {
+            flash_id = FLASH_AREA_IMAGE_SECONDARY(RSS_FIRMWARE_AP_BL2_ID);
+        }
+
+        if(id == flash_id) {
+            flash_map[i].fa_off = (uint32_t)fip_entry->start + FIP_SIGNATURE_AREA_SIZE + tfa_offset;
+            flash_map[i].fa_size = (uint32_t)fip_entry->length;
+        }
+    }
+
+    return true;
+}
+
 /* Read Chip ID and store it in a global variable */
 static void read_chip_id()
 {
@@ -1049,6 +1103,8 @@ static int boot_platform_pre_load_ap_bl2(void)
 {
     enum atu_error_t atu_err;
     int32_t result;
+    uint8_t boot_index = 0;
+    enum fwu_agent_error_t fwu_ret;

     BOOT_LOG_INF("BL2: AP BL2 pre load start");

@@ -1087,6 +1143,16 @@ static int boot_platform_pre_load_ap_bl2(void)
         return 1;
     }
     partition_init(PLATFORM_GPT_IMAGE_AP);
+    fwu_ret = fwu_metadata_provision_ap();
+    if (fwu_ret != FWU_AGENT_SUCCESS) {
+        return 1;
+    }
+
+    boot_index = bl2_get_boot_bank();
+    if (!fill_secure_flash_map_with_data(boot_index)) {
+        BOOT_LOG_ERR("Filling flash map with signed images has failed!");
+        return 1;
+    }
     BOOT_LOG_INF("BL2: AP BL2 pre load complete");

     return 0;
diff --git a/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c
index c3e6b27f8..dc0ee9585 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/flash_map_bl2.c
@@ -16,10 +16,12 @@

 /* Flash device names must be specified by target */
 extern ARM_DRIVER_FLASH FLASH_DEV_NAME;
+extern ARM_DRIVER_FLASH AP_FLASH_DEV_NAME;

 /* Default Drivers list */
 const ARM_DRIVER_FLASH *flash_driver[] = {
     &FLASH_DEV_NAME,
+    &AP_FLASH_DEV_NAME,
 };
 const int flash_driver_entry_num = ARRAY_SIZE(flash_driver);

@@ -123,18 +125,18 @@ struct flash_area flash_map[] = {
         .fa_size = FLASH_AREA_15_SIZE,
     },
     {
-        .fa_id = FLASH_AREA_16_ID,
+        .fa_id = AP_FLASH_AREA_0_ID,
         .fa_device_id = FLASH_DEVICE_ID,
-        .fa_driver = &FLASH_DEV_NAME,
-        .fa_off = FLASH_AREA_16_OFFSET,
-        .fa_size = FLASH_AREA_16_SIZE,
+        .fa_driver = &AP_FLASH_DEV_NAME,
+        .fa_off = AP_FLASH_AREA_0_OFFSET,
+        .fa_size = AP_FLASH_AREA_0_SIZE,
     },
     {
-        .fa_id = FLASH_AREA_17_ID,
+        .fa_id = AP_FLASH_AREA_1_ID,
         .fa_device_id = FLASH_DEVICE_ID,
-        .fa_driver = &FLASH_DEV_NAME,
-        .fa_off = FLASH_AREA_17_OFFSET,
-        .fa_size = FLASH_AREA_17_SIZE,
+        .fa_driver = &AP_FLASH_DEV_NAME,
+        .fa_off = AP_FLASH_AREA_1_OFFSET,
+        .fa_size = AP_FLASH_AREA_1_SIZE,
     },
 };

diff --git a/platform/ext/target/arm/rss/kronos/flash_layout.h b/platform/ext/target/arm/rss/kronos/flash_layout.h
index ea4788a9d..0ceca3030 100644
--- a/platform/ext/target/arm/rss/kronos/flash_layout.h
+++ b/platform/ext/target/arm/rss/kronos/flash_layout.h
@@ -117,39 +117,38 @@
 #define FLASH_AREA_9_ID            (FLASH_AREA_8_ID + 1)
 #define FLASH_AREA_9_OFFSET        (FLASH_AREA_8_OFFSET + FLASH_AREA_8_SIZE)
 #define FLASH_AREA_9_SIZE          (FLASH_LCP_PARTITION_SIZE)
-/* AP BL2 image primary slot */
+/* SI0 image primary slot */
 #define FLASH_AREA_10_ID            (FLASH_AREA_9_ID + 1)
 #define FLASH_AREA_10_OFFSET        (FLASH_AREA_9_OFFSET + FLASH_AREA_9_SIZE)
-#define FLASH_AREA_10_SIZE          (FLASH_AP_BL2_PARTITION_SIZE)
-/* AP BL2 image secondary slot */
+#define FLASH_AREA_10_SIZE          (FLASH_SI_CL0_PARTITION_SIZE)
+/* SI0 image secondary slot */
 #define FLASH_AREA_11_ID            (FLASH_AREA_10_ID + 1)
 #define FLASH_AREA_11_OFFSET        (FLASH_AREA_10_OFFSET + FLASH_AREA_10_SIZE)
-#define FLASH_AREA_11_SIZE          (FLASH_AP_BL2_PARTITION_SIZE)
-/* SI0 image primary slot */
+#define FLASH_AREA_11_SIZE          (FLASH_SI_CL0_PARTITION_SIZE)
+/* SI1 image primary slot */
 #define FLASH_AREA_12_ID            (FLASH_AREA_11_ID + 1)
 #define FLASH_AREA_12_OFFSET        (FLASH_AREA_11_OFFSET + FLASH_AREA_11_SIZE)
-#define FLASH_AREA_12_SIZE          (FLASH_SI_CL0_PARTITION_SIZE)
-/* SI0 image secondary slot */
+#define FLASH_AREA_12_SIZE          (FLASH_SI_CL1_PARTITION_SIZE)
+/* SI1 image secondary slot */
 #define FLASH_AREA_13_ID            (FLASH_AREA_12_ID + 1)
 #define FLASH_AREA_13_OFFSET        (FLASH_AREA_12_OFFSET + FLASH_AREA_12_SIZE)
-#define FLASH_AREA_13_SIZE          (FLASH_SI_CL0_PARTITION_SIZE)
-/* SI1 image primary slot */
+#define FLASH_AREA_13_SIZE          (FLASH_SI_CL1_PARTITION_SIZE)
+/* SI2 image primary slot */
 #define FLASH_AREA_14_ID            (FLASH_AREA_13_ID + 1)
 #define FLASH_AREA_14_OFFSET        (FLASH_AREA_13_OFFSET + FLASH_AREA_13_SIZE)
-#define FLASH_AREA_14_SIZE          (FLASH_SI_CL1_PARTITION_SIZE)
-/* SI1 image secondary slot */
+#define FLASH_AREA_14_SIZE          (FLASH_SI_CL2_PARTITION_SIZE)
+/* SI2 image secondary slot */
 #define FLASH_AREA_15_ID            (FLASH_AREA_14_ID + 1)
 #define FLASH_AREA_15_OFFSET        (FLASH_AREA_14_OFFSET + FLASH_AREA_14_SIZE)
-#define FLASH_AREA_15_SIZE          (FLASH_SI_CL1_PARTITION_SIZE)
-/* SI2 image primary slot */
-#define FLASH_AREA_16_ID            (FLASH_AREA_15_ID + 1)
-#define FLASH_AREA_16_OFFSET        (FLASH_AREA_15_OFFSET + FLASH_AREA_15_SIZE)
-#define FLASH_AREA_16_SIZE          (FLASH_SI_CL2_PARTITION_SIZE)
-/* SI2 image secondary slot */
-#define FLASH_AREA_17_ID            (FLASH_AREA_16_ID + 1)
-#define FLASH_AREA_17_OFFSET        (FLASH_AREA_16_OFFSET + FLASH_AREA_16_SIZE)
-#define FLASH_AREA_17_SIZE          (FLASH_SI_CL2_PARTITION_SIZE)
-
+#define FLASH_AREA_15_SIZE          (FLASH_SI_CL2_PARTITION_SIZE)
+/* FIP image primary slot, in AP secure flash */
+#define AP_FLASH_AREA_0_ID          (FLASH_AREA_15_ID + 1)
+#define AP_FLASH_AREA_0_OFFSET      (0)
+#define AP_FLASH_AREA_0_SIZE        (FLASH_AP_BL2_PARTITION_SIZE)
+/* FIP image secondary slot, in AP secure flash */
+#define AP_FLASH_AREA_1_ID          (AP_FLASH_AREA_0_ID + 1)
+#define AP_FLASH_AREA_1_OFFSET      (AP_FLASH_AREA_0_OFFSET + AP_FLASH_AREA_0_SIZE)
+#define AP_FLASH_AREA_1_SIZE        (FLASH_AP_BL2_PARTITION_SIZE)

 /* Maximum number of image sectors supported by the bootloader. */
 #define MCUBOOT_MAX_IMG_SECTORS    (FLASH_MAX_PARTITION_SIZE / \
@@ -165,20 +164,20 @@
          ((x) == RSS_FIRMWARE_NON_SECURE_ID) ? FLASH_AREA_3_ID : \
          ((x) == RSS_FIRMWARE_SCP_ID)        ? FLASH_AREA_6_ID : \
          ((x) == RSS_FIRMWARE_LCP_ID)        ? FLASH_AREA_8_ID : \
-         ((x) == RSS_FIRMWARE_AP_BL2_ID)     ? FLASH_AREA_10_ID : \
-         ((x) == RSS_FIRMWARE_SI_CL0_ID)     ? FLASH_AREA_12_ID : \
-         ((x) == RSS_FIRMWARE_SI_CL1_ID)     ? FLASH_AREA_14_ID : \
-         ((x) == RSS_FIRMWARE_SI_CL2_ID)     ? FLASH_AREA_16_ID : \
+         ((x) == RSS_FIRMWARE_SI_CL0_ID)     ? FLASH_AREA_10_ID : \
+         ((x) == RSS_FIRMWARE_SI_CL1_ID)     ? FLASH_AREA_12_ID : \
+         ((x) == RSS_FIRMWARE_SI_CL2_ID)     ? FLASH_AREA_14_ID : \
+         ((x) == RSS_FIRMWARE_AP_BL2_ID)     ? AP_FLASH_AREA_0_ID : \
                                               255)
 #define FLASH_AREA_IMAGE_SECONDARY(x) \
         (((x) == RSS_FIRMWARE_SECURE_ID)     ? FLASH_AREA_4_ID : \
          ((x) == RSS_FIRMWARE_NON_SECURE_ID) ? FLASH_AREA_5_ID : \
          ((x) == RSS_FIRMWARE_SCP_ID)        ? FLASH_AREA_7_ID : \
          ((x) == RSS_FIRMWARE_LCP_ID)        ? FLASH_AREA_9_ID : \
-         ((x) == RSS_FIRMWARE_AP_BL2_ID)     ? FLASH_AREA_11_ID : \
-         ((x) == RSS_FIRMWARE_SI_CL0_ID)     ? FLASH_AREA_13_ID : \
-         ((x) == RSS_FIRMWARE_SI_CL1_ID)     ? FLASH_AREA_15_ID : \
-         ((x) == RSS_FIRMWARE_SI_CL2_ID)     ? FLASH_AREA_17_ID : \
+         ((x) == RSS_FIRMWARE_SI_CL0_ID)     ? FLASH_AREA_11_ID : \
+         ((x) == RSS_FIRMWARE_SI_CL1_ID)     ? FLASH_AREA_13_ID : \
+         ((x) == RSS_FIRMWARE_SI_CL2_ID)     ? FLASH_AREA_15_ID : \
+         ((x) == RSS_FIRMWARE_AP_BL2_ID)     ? AP_FLASH_AREA_1_ID : \
                                               255)

 /* Scratch area is not used with RAM loading firmware upgrade */
@@ -218,7 +217,7 @@
 /* Smallest flash programmable unit in bytes */
 #define AP_TFM_HAL_FLASH_PROGRAM_UNIT      (0x1)

-
+#define FIP_SIGNATURE_AREA_SIZE            (0x400)
 /*******************************/
 /* ITS */
 /*******************************/
@@ -263,13 +262,16 @@

 /* FWU Configurations */
 #define NR_OF_FW_BANKS                  (2)
-#define NR_OF_IMAGES_IN_FW_BANK         (7)
+#define NR_OF_IMAGES_IN_FW_BANK         (NR_OF_IMAGES_IN_RSS_FW_BANK + \
+                                         NR_OF_IMAGES_IN_AP_FW_BANK)
 #define NR_OF_IMAGES_IN_RSS_FW_BANK     (7)
+#define NR_OF_IMAGES_IN_AP_FW_BANK      (1)

 /* Space in flash to store metadata and uefi variables */
 #define FWU_METADATA_RSS_FLASH_DEV      (FLASH_DEV_NAME)
 #define FWU_METADATA_FLASH_SECTOR_SIZE  ((0x00001000))

+#define FWU_METADATA_AP_FLASH_DEV       (AP_FLASH_DEV_NAME)
 #define FWU_METADATA_REPLICA_1_OFFSET   (0x5000)  /* 40th LBA */
 #define FWU_METADATA_REPLICA_2_OFFSET   (FWU_METADATA_REPLICA_1_OFFSET + \
                                          FWU_METADATA_FLASH_SECTOR_SIZE)
@@ -287,4 +289,8 @@
 #define BANK_PARTITION_SIZE             (0xFF0000)                 /* ~16MB */
 #define BANK_0_PARTITION_OFFSET         (BL2_BANK_0_OFFSET)
 #define BANK_1_PARTITION_OFFSET         (BL2_BANK_1_OFFSET)
+
+#define FIP_BANK_0_PARTITION_OFFSET     0x7000
+#define FIP_BANK_PARTITION_SIZE         SIZE_DEF_FIP_IMAGE
+#define FIP_BANK_1_PARTITION_OFFSET     (FIP_BANK_0_PARTITION_OFFSET + FIP_BANK_PARTITION_SIZE)
 #endif /* __FLASH_LAYOUT_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c
index 124907cc5..834fed63f 100644
--- a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c
+++ b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c
@@ -22,10 +22,22 @@
 #include "tfm_hal_platform.h"
 #include "uart_stdout.h"

+extern ARM_DRIVER_FLASH FWU_METADATA_AP_FLASH_DEV;
 extern ARM_DRIVER_FLASH FWU_METADATA_RSS_FLASH_DEV;
 extern int is_initialized_rss;
 extern struct fwu_metadata_rss fwu_md_rss;

+#ifndef FWU_METADATA_AP_FLASH_DEV
+    #ifndef FLASH_DEV_NAME
+    #error "FWU_METADATA_AP_FLASH_DEV or AP_FLASH_DEV_NAME must be defined in flash_layout.h"
+    #else
+    #define FWU_METADATA_AP_FLASH_DEV AP_FLASH_DEV_NAME
+    #endif
+#endif
+
+int is_initialized_ap = 0;
+struct fwu_metadata_ap fwu_md_ap;
+
 uint8_t bl2_get_boot_bank(void)
 {
     uint8_t boot_index;
@@ -44,3 +56,60 @@ uint8_t bl2_get_boot_bank(void)
     FWU_LOG_MSG("%s: exit: booting from bank = %u\r\n", __func__, boot_index);
     return boot_index;
 }
+
+enum fwu_agent_error_t fwu_metadata_provision_ap(void)
+{
+    enum fwu_agent_error_t ret;
+    uint32_t image_version = FWU_IMAGE_INITIAL_VERSION;
+
+    ret = fwu_metadata_init(FWU_AP_FLASH_DEV, &is_initialized_ap);
+    if (ret) {
+        return ret;
+    }
+    /* Check by chance if the previous reboot had a firmware data.
+     * If yes, then don't initialize metadata.
+     */
+    metadata_read(FWU_AP_FLASH_DEV, &fwu_md_ap.md);
+    if(fwu_md_ap.md.active_index < 2 || fwu_md_ap.md.previous_active_index <2){
+        if(fwu_md_ap.md.active_index ^ fwu_md_ap.md.previous_active_index)
+            return FWU_AGENT_SUCCESS;
+    }
+    /* Provision FWU Agent Metadata AP secure flash. */
+    memset(&fwu_md_ap, 0, sizeof(struct fwu_metadata_ap));
+
+    fwu_md_ap.md.version = 1;
+    fwu_md_ap.md.active_index = BANK_0;
+    fwu_md_ap.md.previous_active_index = BANK_1;
+
+    /* Bank-0 is the place where images are located at the start
+     * of device lifecycle.
+     */
+    for (uint8_t i = 0; i < NR_OF_IMAGES_IN_AP_FW_BANK; i++) {
+
+        fwu_md_ap.img_entry[i].img_props[BANK_0].accepted = IMAGE_ACCEPTED;
+        fwu_md_ap.img_entry[i].img_props[BANK_0].version = image_version;
+
+        fwu_md_ap.img_entry[i].img_props[BANK_1].accepted = IMAGE_NOT_ACCEPTED;
+        fwu_md_ap.img_entry[i].img_props[BANK_1].version = INVALID_VERSION;
+    }
+
+    /* Calculate CRC32 for fwu metadata. */
+    fwu_md_ap.md.crc_32 = crc32((uint8_t *)&fwu_md_ap.md.version,
+                             sizeof(struct fwu_metadata_ap) - sizeof(uint32_t));
+
+    ret = metadata_write(FWU_AP_FLASH_DEV, &fwu_md_ap.md);
+    if (ret) {
+        return ret;
+    }
+
+    memset(&fwu_md_ap, 0, sizeof(struct fwu_metadata_ap));
+    ret = metadata_read(FWU_AP_FLASH_DEV, &fwu_md_ap.md);
+    if (ret) {
+        return ret;
+    }
+    FWU_LOG_MSG("%s: provisioned values: active = %u, previous = %d\n\r",
+             __func__, fwu_md_ap.md.active_index, fwu_md_ap.md.previous_active_index);
+
+    FWU_LOG_MSG("%s: FWU METADATA PROVISIONED.\n\r", __func__);
+    return FWU_AGENT_SUCCESS;
+}
diff --git a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h
index eea8f87ab..4724e3217 100644
--- a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h
+++ b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h
@@ -123,6 +123,14 @@ struct fwu_metadata_rss {

 } __packed;

+struct fwu_metadata_ap {
+
+        struct metadata md;
+
+        /* Image entry information */
+        struct fwu_image_entry img_entry[NR_OF_IMAGES_IN_AP_FW_BANK];
+
+} __packed;
 enum fwu_agent_error_t {
         FWU_AGENT_SUCCESS = 0,
         FWU_AGENT_ERROR = (-1)
@@ -130,6 +138,7 @@ enum fwu_agent_error_t {

 enum FWU_METADATA_FLASH_DEV {
         FWU_RSS_FLASH_DEV = 0,
+        FWU_AP_FLASH_DEV
 };

 enum fwu_agent_state_t {
diff --git a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent_rss.c b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent_rss.c
index 2d750f1fc..30e025903 100644
--- a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent_rss.c
+++ b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent_rss.c
@@ -32,6 +32,7 @@

 /* Import the CMSIS flash device driver */
 extern ARM_DRIVER_FLASH FWU_METADATA_RSS_FLASH_DEV;
+extern ARM_DRIVER_FLASH FWU_METADATA_AP_FLASH_DEV;

 struct fwu_metadata_rss fwu_md_rss;
 int is_initialized_rss = 0;
@@ -46,6 +47,10 @@ enum fwu_agent_error_t get_fwu_flash_and_img_num(enum FWU_METADATA_FLASH_DEV dev
         *FLASH_DEV = &FWU_METADATA_RSS_FLASH_DEV;
         *img_num = NR_OF_IMAGES_IN_RSS_FW_BANK;
         break;
+    case FWU_AP_FLASH_DEV:
+        *FLASH_DEV = &FWU_METADATA_AP_FLASH_DEV;
+        *img_num = NR_OF_IMAGES_IN_AP_FW_BANK;
+        break;
     default:
         FWU_LOG_MSG("%s: This device is not supported!\n\r", __func__);
         ret = FWU_AGENT_ERROR;
diff --git a/platform/ext/target/arm/rss/kronos/size_defs.h b/platform/ext/target/arm/rss/kronos/size_defs.h
index 4f912204b..d96ced152 100644
--- a/platform/ext/target/arm/rss/kronos/size_defs.h
+++ b/platform/ext/target/arm/rss/kronos/size_defs.h
@@ -29,3 +29,4 @@
 #define SIZE_DEF_SI_CL0_IMAGE       (0x200000) /* 2MB */
 #define SIZE_DEF_SI_CL1_IMAGE       (0x400000) /* 4MB */
 #define SIZE_DEF_SI_CL2_IMAGE       (0x800000) /* 8MB */
+#define SIZE_DEF_FIP_IMAGE          (0x200000) /* 2MB */
--
2.25.1

